uw-frame-components/portal/messages/controllers.js

```diff
@@ -0,0 +1,590 @@
+          // notification
+          if (isHighPriority) {
+            clearPriorityNotificationsFlags();
+          }
+        };
+
+        /**
+         * On-click event to mark a notification as "unseen"
+         * @param notification
+         */
+        vm.restoreNotification = function(notification) {
+          // Remove notification from dismissed array
+          vm.dismissedNotifications = $filter('filterOutMessageWithId')(
+            vm.dismissedNotifications,
+            notification.id
+          );
+
+          // Add notification to non-dismissed array
+          vm.notifications.push(notification);
+
+          // Remove the corresponding entry from
+          // local array of dismissed notification IDs
+          var index = dismissedNotificationIds.indexOf(notification.id);
+          if (index !== -1) {
+            dismissedNotificationIds.splice(index, 1);
+          }
+          // Call service to save changes if k/v store enabled
+          if (SERVICE_LOC.kvURL) {
+            messagesService.setMessagesSeen(dismissedNotificationIds);
+          }
+        };
+
+        /**
+         * Track clicks on "Notifications" links in mobile menu and top bar
+         * @param category
+         * @param action
+         * @param label
+         */
+        vm.pushGAEvent = function(category, action, label) {
+          miscService.pushGAEvent(category, action, label);
+        };
+    }])
+
+    .controller('AnnouncementsController', ['$q', '$log', '$filter',
+      '$sessionStorage', '$scope', '$rootScope', '$document', '$sanitize',
+      '$mdDialog', 'miscService', 'messagesService', 'MESSAGES',
+      function($q, $log, $filter, $sessionStorage, $scope, $rootScope,
+               $document, $sanitize, $mdDialog, miscService, messagesService,
+               MESSAGES) {
+        // //////////////////
+        // Local variables //
+        // //////////////////
+        var vm = this;
+        var allAnnouncements = [];
+        var separatedAnnouncements = {};
+        var seenAnnouncementIds = [];
+        var popups = [];
+
+        // Promise to get seen message IDs
+        var promiseSeenMessageIds = {
+          seenMessageIds: messagesService.getSeenMessageIds(),
+        };
+
+        // ///////////////////
+        // Bindable members //
+        // ///////////////////
+        vm.hover = false;
+        vm.active = false;
+        vm.mode = $scope.mode;
+        vm.announcements = [];
+
+        // //////////////////
+        // Event listeners //
+        // //////////////////
+        /**
+         * When the parent controller has messages, initialize
+         * things dependent on messages
+         */
+        $scope.$watch('$parent.hasMessages', function(hasMessages) {
+          // If the parent scope has messages and notifications are enabled,
+          // complete initialization
+          if (hasMessages && MESSAGES.announcementsEnabled) {
```

Comment:
+1 from me as well. I still wonder about cases where we might want to allow someone to disable/enable all messages (i.e. via a settings toggle), but it would be easy enough to add in down the road.